examples/mhp/stencil-2d-matrix-rows.cpp

```diff
@@ -0,0 +1,181 @@
+
+#include <algorithm>
+
+#include "../include/data-utils.hpp"
+#include "cxxopts.hpp"
+#include "mpi.h"
+
+#include "dr/mhp.hpp"
+
+using T = float;
+
+int comm_rank;
+
+cxxopts::ParseResult options;
+
+std::size_t nc = 0;
+std::size_t nr = 0;
+std::size_t steps = 0;
+
+//
+//  verification
+//
+int matrix_compare(std::string label, dr::mhp::distributed_dense_matrix<T> &dm,
+                   std::vector<std::vector<T>> &vv) {
+  int res = 0;
+  for (std::size_t i = 0; i < vv.size(); i++) {
+    for (std::size_t j = 0; j < vv[i].size(); j++) {
+      if (!is_equal(vv[i][j], (T)dm.begin()[{i, j}])) {
+        fmt::print("{}: {} Fail in cell [{},{}] values: ref {} matrix {})\n",
+                   comm_rank, label, i, j, vv[i][j], (T)dm.begin()[{i, j}]);
+        res = -1;
+      }
+    }
+  }
+  return res;
+}
+
+//
+// calculation formula
+//
+T calculate(T arg1, T arg2, T arg3, T arg4, T arg5) {
+  return arg1 + arg2 + arg3 + arg4 + arg5;
+}
+
+//
+// verification of results
+//
+void stencil_op_verify(std::vector<std::vector<T>> &va,
+                       std::vector<std::vector<T>> &vb) {
+  for (std::size_t i = 1; i < va.size() - 1; i++) {
+    for (std::size_t j = 1; j < va[i].size() - 1; j++)
+      vb[i][j] = calculate(va[i - 1][j], va[i][j], va[i + 1][j], va[i][j - 1],
+                           va[i][j + 1]);
+  }
+}
+
+int stencil_check(dr::mhp::distributed_dense_matrix<T> &a,
+                  dr::mhp::distributed_dense_matrix<T> &b) {
+
+  std::vector<std::vector<T>> va(nr), vb(nr);
+
+  for (auto &r : va) {
+    r.resize(nc);
+    rng::iota(r, 10);
+  }
+
+  for (auto &r : vb) {
+    r.resize(nc);
+    rng::iota(r, 10);
+  }
+
+  for (std::size_t s = 0; s < steps; s++) {
+    stencil_op_verify(va, vb);
+    std::swap(va, vb);
+  }
+
+  return matrix_compare("A", a, (steps % 2) ? vb : va) +
+         matrix_compare("B", b, (steps % 2) ? va : vb);
+}
+
+//
+// stencil
+//
+
+auto stencil_op = [](auto &&p) {
+  std::vector<T> out_row((*p).size());
+
+  out_row[0] = p[0][0];
+  for (std::size_t i = 1; i < nc - 1; i++) {
+    out_row[i] =
+        calculate(p[-1][i], p[0][i - 1], p[0][i], p[0][i + 1], p[1][i]);
+  }
+  out_row[nc - 1] = p[0][nc - 1];
+  return out_row;
+};
+
+int stencil() {
+  dr::mhp::halo_bounds hb(1); // 1 row
+  dr::mhp::distributed_dense_matrix<T> a(nr, nc, -1, hb), b(nr, nc, -1, hb);
+
+  // the same operation on every row
+  dr::mhp::for_each(a.rows(), [](auto row) { rng::iota(row, 10); });
+
+  // different operation on every row is posiible by access to the row (*r)
+  for (auto r = b.rows().begin(); r != b.rows().end(); r++) {
+    if (r.is_local())
+      rng::iota(*r, 10);
+  }
+
+  // subranges of rows
+  auto in = rng::subrange(a.rows().begin() + 1, a.rows().end() - 1);
+  auto out = rng::subrange(b.rows().begin() + 1, b.rows().end() - 1);
+
+  for (std::size_t s = 0; s < steps; s++) {
+    dr::mhp::halo(in).exchange();
+    dr::mhp::transform(in, out.begin(), stencil_op);
```

Comment:
I agree, still I had problems with for_each and zip_view, so that an iterator is passed to the stencil, not ref to an element. dm_rows class requires some more work (see https://github.com/oneapi-src/distributed-ranges/pull/191#discussion_r1235111325)